doc/man3/SSL_CTX_set_cert_comp_cb.pod

```diff
@@ -0,0 +1,131 @@
+ int SSL_set_cert_comp_cb(SSL *ssl, OSSL_cert_comp_fn cert_comp_cb, void *arg);
+
+ unsigned char *SSL_get_cert_to_compress(SSL *ssl, size_t *length);
+ int SSL_set1_compressed_cert(SSL *ssl, uint16_t algorithm, unsigned char *data,
+                              size_t length);
+
+=head1 DESCRIPTION
+
+These functions control the certificate compression feature. Certificate
+compression is only available for TLSv1.3 as defined in RFC8879.
+
+SSL_CTX_set_cert_comp_preference() and SSL_set_cert_comp_preference() are used
+to specify the preferred compression algorithms. Only those algorithms enabled
+in the library will be accepted, unknown algorithms are ignored. The B<algs>
+argument is an array of algorithms, and B<num_args> is the number of algorithms.
+The following compression algorithms may be used:
+
+=over 4
+
+=item TLSEXT_comp_cert_brotli
+
+=item TLSEXT_comp_cert_zlib
+
+=item TLSEXT_comp_cert_zstd
+
+=back
+
+The above is also the default preference order.
+
+SSL_CTX_set_cert_comp_cb() and SSL_set_cert_comp_cb() set the certificate
+compression callback, defined as type I<OSSL_cert_comp_fn>. If sending compressed
+certificates is not disabled via SSL_OP_NO_TX_CERTIFICATE_COMPRESSION, and the
+TLSv1.3 peer has offered a supported certificate compression extension, the
+B<cert_comp_cb> callback is invoked with the B<arg>.
+
+The certificate compression callback must return one of the following:
+
+=over 4
+
+=item Z<>0
+
+No certificate compression is desired; the state machine will send a regular
+Certificate message.
+
+=item E<gt>0
+
+Use the compressed certificate specified by SSL_set1_compressed_cert(). If
+SSL_set1_compressed_cert() has not been called, the internal TLS state machine
+will compress the certificate.
+
+=item E<lt>0
+
+An error has occured; terminate the connection.
+
+=back
+
+SSL_get_cert_to_compress() is called by the certificate compression callback
+to get the certificate in a ready-to-compress format (i.e. it includes the
+certificate context and certificate chain). The return value is allocated and
+must be freed by the caller.
+
+SSL_set1_compressed_cert() is called by the certificate compression callback
+to set the now-compressed certificate into the state machine. After setting
+the certificate, the certificate compression callback must return >0 to use
+the now-compressed certificate.
+
+=head1 NOTES
+
+Sending compressed certificates may be disabled on a connection via the
+SSL_OP_NO_TX_CERTIFICATE_COMPRESSION option. Receiving compressed certificates
+may be disabled on a connection via the SSL_OP_NO_RX_CERTIFICATE_COMPRESSION
+option.
+
+If, after calling SSL_get_cert_to_compress() and compressing the
+certificate, the data expands, the application can discard the results and
+return 0 from the certificate compression callback.
+
+Since servers do not have a certificate context, a server may pre-compress
+its certificate and give it to SSL_set1_compressed_cert() during the callback.
```

Comment:
I would argue that doing it automatically is potentially pointless and wasteful. I much prefer it to be under the control of the user.

There are a few nuances.

Number of supported compression algorithms. There are three, and I don't think we'd want to waste time and space compressing for each algorithm. The preference options does allow us to limit this, but doing it automatically? (Admittedly, Chrome just supports `brotli` the last time I checked.)

There's the case where a provider already has the certificate compressed (from the same source it retrieved the certificate - I could imagine CDN providers doing this), this means compression has already been done. So perhaps better for the the API to allow setting the compressed certificate into the `SSL` or `SSL_CTX` along with the certificate? (This gets into the morass of updating certificate, key and compressed certificate in the correct order, and potentially having to decompress a certificate to ensure that it matches the certificate already set.)

Or a pre-compress API to cause compression to be directed by the user before connection (e.g. `SSL_CTX_compress_certificate(SSL_CTX*, int alg))

The callback API allows for this caching and avoiding unnecessary compressions and storage space. That being said, TLSv1.3 is better than TLSv1.2, in that you're unlikely to have both RSA and ECDSA certificates simultaneously, but an `SSL_CTX` could support both protocol versions, and subsequently have multiple certificates. Multiply by the number of algorithms...

That being said I *do* want to keep the callback API, it is useful as described above. I suppose the compressed cert  could be cached into the `SSL_CTX` (little point in caching inside the `SSL`).

Not saying that this PR should be approved without a caching mechanism, but I'd like time to think about it. (I'm leaning towards caching in the `SSL_CTX` along with the pre-compress API.)

Reply:
I guess my concern is with turning this on by default if we don't have an easy mechanism for caching the certificate.

Perhaps another way to approach this is that the user must call `SSL_CTX_compress_certificate(SSL_CTX *ctx, int alg)` before hand (on the server). Only if we *have* a compressed certificate for a given compression algorithm already (for all certificates that we have enabled) would we add that algorithm to our available list (so we wouldn't negotiate it if we don't have one). We can override that behaviour in the presence of a callback to give advanced users more control.

Alternative to `SSL_CTX_compress_certificate`, you could imagine an `SSL_CTX_use_certificate_ex` API which works the same way as `SSL_CTX_use_certificate` but additionally gives you a mechanism to say whether you want to compress the certificate at the same time, and if so with which algorithms.

Reply:
I like the `SSL_CTX_compress_cert` API, as it will compress the existing configured cert safely (i.e. little chance of the wrong data being used).

As of the latest set of changes, the configured certificate can only be pre-compressed with a single algorithm. If that algorithm is not the only preferred algorithm (e.g. because the preference list was left at the default), then the connection will automatically compress as needed.
e.g. built with brotli and zlib. Precompress with brotli, but client only claims support for zlib, so the state machine will compress with zlib. (I should write a test...)
In this case, the just-compressed certificate is not cached; I don't really want to push the certificate up into the SSL_CTX (given `SSL_use_certificate()`), and because it's common use to share an SSL_CTX across multiple threads (treating the SSL_CTX as read-only, except for session cache).
So, the SSL can get the certificate from the SSL_CTX, but not the other way around.